### PR TITLE
gh-105375: Improve PyErr_WarnExplicit() error handling

### PR DESCRIPTION
**Pull request by erlend-aasland**
Bail on first error to prevent exceptions from possibly being
overwritten.

* Issue: gh-105375

### CODE REVIEW

File: Python/_warnings.c [hunk 1]
```diff
@@ -1298,12 +1298,16 @@ PyErr_WarnExplicit(PyObject *category, const char *text,
...
     PyObject *message = PyUnicode_FromString(text);
+    if (message == NULL) {
+        goto exit;
+    }
     PyObject *filename = PyUnicode_DecodeFSDefault(filename_str);
+    if (filename == NULL) {
+        goto exit;
+    }
     PyObject *module = NULL;
     int ret = -1;
```
Comment (sobolevn) [inline Python/_warnings.c]: You can move this line to 1300 to fix the warning :)

Comment (erlend-aasland) [reply Python/_warnings.c]: Oh, thanks! :) I just removed the variable; IMO that makes the code easier to read.

Comment (erlend-aasland) [reply Python/_warnings.c]: I ended up rewriting the whole PR.

**Review by sobolevn — APPROVED** [review]
Great, thank you!

**Comment by erlend-aasland** [comment]
> Great, thank you!

Thanks for all the reviews! Very helpful.

**Comment by erlend-aasland** [comment]
Ref. leak runs are still unusable bco. the `test_import` and `test_peg_generator` failures :(

**Pull request merged**